rfcs/rdm-0064-custom-fields.md

```diff
@@ -0,0 +1,558 @@
+
+```yaml
+id: ...
+icon: ...
+props: ...
+description: ...
+tags: ...
+title: ...
+```
+
+- Fields are configured globally for all records and for all communities. Note that they do not mix, those appearing in records will not appear in communities. However, all communities will have the same fields.
+- You can hide (not display) fields in the UI. However, they will be available in the REST API. Moreover, hiding them will hide them globally. This means that the upload or settings form will not display them, but they will be visible in the landing page (e.g. if submitted via REST API).
+- There are no per-field permissions checks on these fields. They do respect the records permissions (i.e. will not show if the record is restricted).
+- It is not possible to configure the fields using the UI, it will be done in the instance `invenio.cfg` file.
+- Compatibility with OpenSearch has not been tested. Only Elasticsearch 7 has been tested.
+- In the forms, you can add fields to new groups but not to existing ones (e.g. Basic Information).
+- Translations (i18n) is only supported for labels and descriptions of all fields. Vocabularies also support translation of the values as long as those have been added when imported.
+
+## Detailed design
+
+An instance administrator can add custom fields. **Updating and deleting a custom field is not possible**. This constraint is imposed to avoid data migration and potential inconsistencies.
+
+### Configuring a _custom field_
+
+The first step is to define how custom fields can be added/configured. A top to bottom approach (from more to less abstracted) will be followed to explain the configuration. Taking the following use case:
+
+_At CERN, I want to choose the experiment to which the record belong to, assuming experiments come from a controlled vocabulary. In addition, I want to be able to filter the record search by experiment._
+
+The instance administrator will change the configuraiton variables `RDM_RECORDS_CUSTOM_FIELDS` and `RDM_RECORDS_CUSTOM_FIELDS_UI` in the `invenio.cfg` configuration file. The following example shows only the configuration needed to **enable the custom field** in the data model not how to display them. This second part will be explained in the [_Adding-a-custom-UI-widget_](#Add a custom UI widget) section.
+
+```python
+RDM_RECORDS_CUSTOM_FIELDS = {
+    "experiment": VocabularyCF(
+        name="experiment",
+        vocabulary_id="cernexperiments",
+    ),
+}
+```
+
+> Note: In the above configuration and along this RFC configuration variables are prefixed with `RDM_RECORDS_`, this means that custom fields are being added to _records_. However, the same configuration variables are available in communities and will be used then adding custom fields to them. Namely, `COMMUNITIES_CUSTOM_FIELDS` and `COMMUNITIES_CUSTOM_FIELDS_UI`.
```

Comment:
Agree on this point. There is an issue to investigate if there is a better way than exposing this "cumbersome" double config (e.g. records and communities). What about `RECORDS_CUSTOM_FIELDS` and `COMMUNITIES_CUSTOM_FIELDS`? It would clash with `invenio-records`? I'm mainly not convinced of giving the full `RDM_` prefix to something that is just records, since communities is also part of "RDM". Facets in the end is a generic, although it might be the same when/if configuring communities search facets. Will change it for now, then maybe there is another way found when tcakling the specific issue.

Reply:
https://github.com/inveniosoftware/invenio-rdm-records/issues/1006